### PR TITLE
Optimize performance of log_softmax

### PR DESCRIPTION
**Pull request by ZzSean**
### PR types

Performance optimization
### PR changes

OPs
### Describe

Optimize performance of log_softmax by the unified logic of softmax

| case                         | pytorch | paddle 优化前 | diff          | paddle 优化后 | diff          | 加速比 |
| ---------------------------- | ------- | ------------- | ------------- | ------------- | ------------- | ------ |
| fp32,[16,1000] axis:-1       | 0.00933 | 0.01081       | 差于 (15.86%) | 0.00774       | 优于 (20.54%) | 1.40   |
| fp16,[16,1000] axis:-1       | 0.01043 | 0.01507       | 差于 (44.49%) | 0.00938       | 优于 (11.19%) | 1.61   |
| fp32,[32,12,128,128] axis:-1 | 0.15398 | 0.15366       | 打平 (0.21%)  | 0.15498       | 打平 (0.65%)  | 0.99   |
| fp16,[32,12,128,128] axis:-1 | 0.08158 | 0.08094       | 打平 (0.79%)  | 0.08594       | 差于 (5.34%)  | 0.94   |
| fp32,[15,16,33,33] axis:-1   | 0.00818 | 0.00995       | 差于 (21.64%) | 0.00968       | 差于 (18.34%) | 1.03   |
| fp16,[15,16,33,33] axis:-1   | 0.00818 | 0.01000       | 差于 (22.25%) | 0.00980       | 差于 (19.80%) | 1.02   |
| fp32,[128,128,16,16] axis:0  | 0.32817 | 1.16099       | 差于 (2.54x)  | 0.14324       | 优于 (1.29x)  | 8.11   |
| fp16,[128,128,16,16] axis:0  | 0.30698 | 0.78857       | 差于 (1.57x)  | 0.11444       | 优于 (1.68x)  | 6.89   |
| fp32,[512,896,4,12] axis:1   | 1.43943 | 15.33653      | 差于 (9.65x)  | 1.05637       | 优于 (36.26%) | 14.52  |
| fp16,[512,896,4,12] axis:1   | 1.13875 | 14.92736      | 差于 (12.11x) | 0.71325       | 优于 (59.66%) | 20.93  |

### CODE REVIEW

**Review by Avin0323 — DISMISSED** [review]
LGTM for unity_build_rule.cmake

File: paddle/fluid/operators/log_softmax_op.cu [hunk 1]
```diff
@@ -483,3 +516,35 @@ REGISTER_OP_CUDA_KERNEL(
...
+#else
+REGISTER_OP_KERNEL(log_softmax, CUDNN, plat::CUDAPlace,
+                   ops::LogSoftmaxCUDNNKernel<float>,
+                   ops::LogSoftmaxCUDNNKernel<double>,
+                   ops::LogSoftmaxCUDNNKernel<plat::float16>);
+REGISTER_OP_KERNEL(log_softmax_grad, CUDNN, plat::CUDAPlace,
+                   ops::LogSoftmaxGradCUDNNKernel<float>,
+                   ops::LogSoftmaxGradCUDNNKernel<double>,
+                   ops::LogSoftmaxGradCUDNNKernel<plat::float16>);
+#endif
```
Comment (AshburnLee) [inline paddle/fluid/operators/log_softmax_op.cu]: 这三部分注册可以通过可变参数宏优化。

**Review by AshburnLee — DISMISSED** [review]
LGTM.

**Review by limin2021 — DISMISSED** [review]
LGTM.

File: paddle/fluid/operators/log_softmax_op.cc [hunk 2]
```diff
@@ -56,6 +65,11 @@ class LogSoftmaxOpMaker : public framework::OpProtoAndCheckerMaker {
                  "The dimension index of Input(x) to perform log_softmax,"
                  "default -1 for last dimension")
         .SetDefault(-1);
+    AddAttr<bool>(
```
Comment (Xreki) [inline paddle/fluid/operators/log_softmax_op.cc]: 不建议新增属性，直接改原CUDA Kernel吧。

Comment (ZzSean) [reply paddle/fluid/operators/log_softmax_op.cc]: done，thx

File: paddle/phi/kernels/gpudnn/softmax_gpudnn.h [hunk 3]
```diff
@@ -864,6 +904,32 @@ static bool CanUseCudnnSoftmax(const GPUContext& dev_ctx) {
...
+#if CUDNN_VERSION < 8100
+template <>
+inline void SoftmaxForwardCudnnKernel<phi::dtype::bfloat16>(
+    const GPUContext& dev_ctx,
+    const DenseTensor& x,
+    const int axis,
+    const bool log_mode,
+    DenseTensor* out) {
+  PADDLE_THROW(
+      "This kernel is not supported when the dtype is bf16 and CUDNN_VERSION < "
```
Comment (Xreki) [inline paddle/phi/kernels/gpudnn/softmax_gpudnn.h:916]: `PADDLE_THROW`也要指明`error`类型。

Comment (ZzSean) [reply paddle/phi/kernels/gpudnn/softmax_gpudnn.h:916]: done, thx

**Review by Xreki — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]
LGTM for PADDLE_ENFORCE
